# EpitopeScan
EpitopeScan is a Python3 toolset facilitating the tracking of mutations in SARS-CoV-2 immunogenic epitopes.

**Contents:**
1. Repository contents
2. Download and set-up
3. EpitopeScan usage
	3.1
	3.2 
	3.3
4. Algorithm details

## 1. Repository contents

- `EpitopeScan.py` is a command line tool for mutation report from genome data and calculation of general mutation statistics.

- `EpitopeScanGUI.py` is a Graphical User Interface (GUI) application. It uses system's default web-brwoser to provide interactive interface for analysis of mutation data generated by `EpitopeScan.py`  for samples supported by metadata on date and lineage.

- `utils` folder contains supporting Python3 code with functions and classes utilised by command line and GUI tools.

- `reference_sequences` contains reference SARS-CoV-2 data listed from [GISAID](https://gisaid.org/wiv04/), which is reference genome `EPI_ISL_402124.fasta`, table with Open Reading Frames information `ORFs_reference.txt` (ORF name, genome start and end coordinates, name of translated protein, translated protein length), reference protein sequences `protein_sequences_reference.fasta`.

- `test` contains test data and a script `run_EpitopeScan_test.py` for the command-line application testing.

- `requirements.txt` is a list of packages to be installed when setting up Python3 environment

## 2.  Download and set-up

Using terminal create a new Python3 environment and install dependancies using `venv` and `pip` or `conda`:

```
# create and activate environment using venv
python3 -m venv EpitopeScan_env
source EpitopeScan_env/bin/activate

# or create and activate environment using conda
conda create --name EpitopeScan_env
conda activate EpitopeScan_env

# then download required packages using pip
pip install -r requirements.txt

# or using conda
conda install -c bioconda --file requirements.txt
```
More tutorials on [pip and virtual environments](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/) and [conda](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html).

Clone EpitopeScan repository to your local system by running: 

```
git clone git@github.com:Aleksandr-biochem/EpitopeScan.git
```

Prior to starting the package usage you may run test script a described in [section 3.4](#Running test scripts) to verify correct performance of EpitopeScan command line tool.

If you want to quckly call EpitopeScan from anywhere on your system, consult these tutorials to add EpitopeScan folder to system paths on [Windows](https://correlated.kayako.com/article/40-running-python-scripts-from-anywhere-under-windows), [Mac](https://stackoverflow.com/questions/22465332/setting-path-environment-variable-in-osx-permanently), [Linux](https://www.geeksforgeeks.org/run-python-script-from-anywhere-in-linux/). 

## 3. EpitopeScan usage

### 3.1 Input data

**EpitopeScan requires the following inputs:**

a. *Peptide sequences* for analysis. Multiple sequences may be provided as FASTA file (see usage examples below).

b. *Genome Multiple Sequence Alignment (MSA)* in FASTA format. EpitopeScan was originally developed to analyse precalculated genome alignments from [COG-UK](https://www.cogconsortium.uk/priority-areas/data-linkage-analysis/public-data-analysis/). The tool does not perform genome alignments. If you want to prepare your set of genomes for analysis, you can consult tutorial and use `EPI_ISL_402124.fasta` as the reference sequence.

c. *Sample metadata*  on samples dates and lineages in CSV format. MSa analysis can be performed without this input. However, it will only be possible to count mutations without any account for sampling date and lineage and the majority of EpitopeScan GUI report will be irrelevant. EpitopeScan was originally configured to deal with metadata from [COG-UK](https://www.cogconsortium.uk/priority-areas/data-linkage-analysis/public-data-analysis/). If you wish to utilise your custom table, make sure that this table contains the following columns: sequence\_name, sample\_date, epi\_week, lineage.

### 3.2 Running command line tool

`EpitopeScan.py` incorporates two modes:
- *scan* to perform MSa file analysis and generate mutation data
- *stat* to generate brief mutation statistics from preexisting mutation data 

The following execution examples are given under assumption of running the tools from `EpitopeScan` folder as `./EpitopeScan_tool_name.py`. In general the tools should be called as `current/path/to/EpitopeScan_tool_name.py`.

#### 3.2.1 Scan mode

The *scan* mode accepts epitope(s), genome MSA for analysis and metadata files to generate mutation data. Terminal log messages track run configurations and process, including how epitope(s) are mapped onto reference genome and how many genomes are processed at the moment. Finally the mutation summary is generated into terminal stdout including the number of mutations for each analysed epitope and the list of discovered mutations with their counts and BLOSUM scores.
There is a help section to navigate flags:

```
./EpitopeScan.py scan -h
```

Flag description:
- `-e` (--epitope) Input peptide epitope name and sequence, comma separated. E.x. S1,VGYWA
- `-f` (--file) For multiple input peptides, provide a path to fasta file with multiple input peptides
- `--msa` Path to input MSA fasta file
- `--metadata` Metadata .csv file to bind with mutaion data
- `-o` (--out) Output directory name
- `-t` (--tag) Sample tag to filter
- `-q` (--quality\_filter) Threshold of max N bases proportion in genome. Recommended 0.05
- `-n` (--no\_ambiguity) Treat any presence ambiguous bases in peptide region as insufficient coverage
- `-b` (--blosum) BLOSUM version for mutation scoring. Default 90
- `-s` (--sort), options: {0,1}, Sort mutations summary by count(0) or score(1). Default 0
- `-a` (--stat), options: {0,1}, Stat individual mutations (0) or combinations(1).  Default 0
- `--stat_with_metadata`  Only stat samples with metadata

Basic *scan* run to generate mutation data for 1 peptide into automatically named folder:

```
./EpitopeScan.py scan 
```

Perform analysis for multiple peptodes from FASTA file and combine mutation data with samples metada:

```
./EpitopeScan.py scan 
```

Add genome quality threshold of 0.07 (7%) and mark any sample with ambiguous bases in epitope's region as insufficient coverage sample:

```
./EpitopeScan.py scan 
```

#### 3.2.2 Stat mode

The *stat* mode accepts preexisting *scan* output and generates summary report into terminal stdout under specified options.
There is a help section to navigate flags:

```
./EpitopeScan.py stat -h
```

Flag description:
- `-i` (--input) Direcory with scan output
- `-b` (--blosum) BLOSUM version for mutation scoring. Default 90
- `-s` (--sort), options: {0,1}, Sort mutations summary by count(0) or score(1). Default 0
- `-a` (--stat), options: {0,1}, Stat individual mutations (0) or combinations(1).  Default 0
- `--stat_with_metadata`  Only stat samples with metadata
- `--start\_date` Subset after this date, dd/mm/yyyy
- `--end\_date` Subset before this date, dd/mm/yyyy

Basic run can be launched as follows:

```
./EpitopeScan.py stat 
```

To stat unique combinations instead of individual mutations in a desired date range, and sort summary table by score:

```
./EpitopeScan.py stat 
```

#### 3.2.3 Mutation data output

For each peptide 3 tables are generated and named after following templates:
- `EpitopeName\_mutation\_data.tsv` sontains sequence\_name, NA\_mutations, AA\_mutations columns followed by one-hot-encoding columns detailing presence of discovered AA mutations in each sample
- `EpitopeName\_AA\_mutation\_matrix.csv` count matrix with index corresponding to AA residue in reference epitope sequence and the columns presenting accurance for each residue and deletion (Î”) at this position
- `EpitopeName\_NA\_mutation\_matrix.csv` likewise for reference nucleic acid sequece coding the epitope

When multiple peptides provided the output foler will contain separate subfolders with tables for each peptide.

### 3.3 Running graphical user-interface tool

EpitopeScan GUI runs locally using the default web-browser as an interface. To launch the application execute:

```
streamlit run ./EpitopeScanGUI.py --server.maxUploadSize 1000
```

Upload mutation data files for analysis and explore interactive summary.
Note that by default streamlit limits file upload to 200 MB therefore it is recommended to launch the app specifying the larger limit through `--server.maxUploadSize`.

### 3.4 Running test scripts

Folder `test` contains `run_EpitopeScan_test.py` and test data. This script will perform analysis for each epitope and compare EpitopeScan output to reference mutation data.
Run the test script after installing EpitopeScan to verify the correct performance of the tool:

```
test/run_EpitopeScan_test.py
```

## 4. Algorithm details

This section contains several remarks about EpitopeScan algorithm, which may be important to correctly interpret the output:

